gostatsd.go

```diff
@@ -23,35 +25,104 @@ var (
 )
 
 func main() {
-	runtime.GOMAXPROCS(runtime.NumCPU())
 	s := statsd.NewServer()
-	s.AddFlags(pflag.CommandLine)
-	pflag.Parse()
-
-	if s.CPUProfile != "" {
-		f, err := os.Create(s.CPUProfile)
+	err, version, CPUProfile := setConfiguration(s)
+	if err != nil {
+		log.Fatalf("Error while parsing configuration: %v\n", err)
+	}
+	if version {
+		fmt.Printf("Version: %s - Commit: %s - Date: %s\n", Version, GitCommit, BuildDate)
+		return
+	}
+	exitCode := 0
+	defer func(e *int) {
+		//pprof.Lookup("goroutine").WriteTo(os.Stdout, 1)
+		if *e != 0 {
+			os.Exit(*e)
+		}
+	}(&exitCode)
+	if CPUProfile != "" {
+		f, err := os.Create(CPUProfile)
 		if err != nil {
-			log.Fatalf("%v", err)
+			log.Fatalf("Failed to open profile file: %v", err)
 		}
+		defer func() {
+			if err := f.Close(); err != nil {
+				log.Errorf("Failed to close profile file: %v", err)
+			}
+		}()
 		pprof.StartCPUProfile(f)
 		defer pprof.StopCPUProfile()
+	}
 
-		c := make(chan os.Signal, 1)
-		signal.Notify(c, os.Interrupt)
-		go func() {
-			for range c {
-				pprof.StopCPUProfile()
-				os.Exit(0)
-			}
-		}()
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	defer cancelFunc()
+	cancelOnInterrupt(cancelFunc)
+
+	log.Info("Starting server")
+
+	if err := s.Run(ctx); err != nil {
+		if err != context.Canceled {
+			exitCode = 1
+			log.Errorf("%v", err)
+		}
 	}
+}
 
-	if s.Version {
-		fmt.Printf("Version: %s - Commit: %s - Date: %s\n", Version, GitCommit, BuildDate)
-		os.Exit(0)
+// cancelOnInterrupt calls f when os.Interrupt or SIGTERM is received
+func cancelOnInterrupt(f context.CancelFunc) {
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
+	go func() {
+		select {
+		case <-c:
+			f()
+		}
+	}()
+}
+
+// addConfigFlags adds flags supported by Config to the specified FlagSet
+func addConfigFlags(s *statsd.Server, fs *pflag.FlagSet) {
```

Comment:
Ok, yes, I understand the intent. Let me explain my point of view. `statsd` package should be designed (as any piece of useful software) to be reusable. It is nice when you can reuse just the bit that you need, without the "batteries included" things. IMHO this method might be considered as specific to this particular program - it parses options from command line/etc; it is part of the bootstrap process. Reusable packages expose an API. It should be as clean and small as possible. "Perfect" configuration API is when you just inject the configuration values and that is it, nothing else it needed. I agree that in some cases this method may be reused as well, but if I was to write some other console application that also used `statsd` package I would probably write a new command line parsing function because I might have different flag names (e.g. I might not need all of them by design), sources of configuration and help messages. Does it make sense? WDYT?